pkg/controller/networkpolicyrecommendation/controller.go

```diff
@@ -0,0 +1,178 @@
+	"antrea.io/theia/pkg/client/clientset/versioned"
+	crdv1a1informers "antrea.io/theia/pkg/client/informers/externalversions/crd/v1alpha1"
+	"antrea.io/theia/pkg/client/listers/crd/v1alpha1"
+)
+
+const (
+	controllerName = "NetworkPolicyRecommendationController"
+	// Set resyncPeriod to 0 to disable resyncing.
+	resyncPeriod time.Duration = 0
+	// How long to wait before retrying the processing of an Service change.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+	// Default number of workers processing an Service change.
+	defaultWorkers = 4
+)
+
+type NPRecommendationController struct {
+	crdClient versioned.Interface
+
+	npRecommendationInformer cache.SharedIndexInformer
+	npRecommendationLister   v1alpha1.NetworkPolicyRecommendationLister
+	npRecommendationSynced   cache.InformerSynced
+	// queue maintains the Service objects that need to be synced.
+	queue workqueue.RateLimitingInterface
+}
+
+func NewNPRecommendationController(
+	crdClient versioned.Interface,
+	npRecommendationInformer crdv1a1informers.NetworkPolicyRecommendationInformer,
+) *NPRecommendationController {
+	c := &NPRecommendationController{
+		crdClient:                crdClient,
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "npRecommendation"),
+		npRecommendationInformer: npRecommendationInformer.Informer(),
+		npRecommendationLister:   npRecommendationInformer.Lister(),
+		npRecommendationSynced:   npRecommendationInformer.Informer().HasSynced,
+	}
+
+	c.npRecommendationInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addNPRecommendation,
+			DeleteFunc: c.deleteNPRecommendation,
+		},
+		resyncPeriod,
+	)
+
+	return c
+}
+
+func (c *NPRecommendationController) addNPRecommendation(obj interface{}) {
+	npReco, _ := obj.(*crdv1alpha1.NetworkPolicyRecommendation)
+	klog.V(2).Infof("Processing NP Recommendation %s ADD event, labels: %v", npReco.Name, npReco.Labels)
+	namespacedName := apimachinerytypes.NamespacedName{
+		Namespace: npReco.Namespace,
+		Name:      npReco.Name,
+	}
+	c.queue.Add(namespacedName)
+}
+
+func (c *NPRecommendationController) deleteNPRecommendation(old interface{}) {
+	npReco, ok := old.(*crdv1alpha1.NetworkPolicyRecommendation)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Error decoding object when deleting NP Recommendation, invalid type: %v", old)
+			return
+		}
+		npReco, ok = tombstone.Obj.(*crdv1alpha1.NetworkPolicyRecommendation)
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting NP Recommendation, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
+	klog.V(2).Infof("Processing NP Recommendation %s DELETE event, labels: %v", npReco.Name, npReco.Labels)
+	namespacedName := apimachinerytypes.NamespacedName{
+		Namespace: npReco.Namespace,
+		Name:      npReco.Name,
+	}
+	c.queue.Add(namespacedName)
```

Comment:
Should it be c.queue.forget(namespacedName)?

Reply:
`c.queue` is the actual worker queue to reconcile the resources. here we're just adding the key to the workqueue to indicate that some resource is removed and reconciliation is needed (i.e. deletion of child res related to this job CR). the key is forgotten only when reconciliation is done for it.